### PR TITLE
Add dataframe visualization

### PR DESCRIPTION
**Pull request by sharanramjee**
## Description

Added visualization support for text and structured datasets by creating dataframes using pandas and exporting them as .csv files.
  
## Checklist
* [X] Address all TODO's
* [X] If using additional dependencies (e.g. `scipy`), use [lazy_imports](https://www.tensorflow.org/datasets/add_dataset#extra_dependencies) (if applicable)
* [X] [Lint](https://www.tensorflow.org/datasets/add_dataset#5_check_your_code_style) code

### CODE REVIEW

File: tensorflow_datasets/core/visualization/as_dataframe.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    **options_kwargs: Additional display options, specific to the dataset type
+      to visualize. Are forwarded to `tfds.visualization.Visualizer.show`.
+      See the `tfds.visualization` for a list of available visualizers.
+
+  Returns:
+    dataframe: The `pandas.DataFrame` object
+  """
+  if not isinstance(ds_info, dataset_info.DatasetInfo):  # Arguments inverted
+    # `absl.logging` does not appear on Colab by default, so uses print instead.
+    print('WARNING: For consistency with `tfds.load`, the `tfds.show_examples` '
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/as_dataframe.py:68]: This is a deprecation warning which should be removed, as this is a new function.

File: tensorflow_datasets/core/visualization/as_dataframe.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+from typing import Any
+
+import tensorflow.compat.v2 as tf
+from tensorflow_datasets.core import dataset_info
+from tensorflow_datasets.core.visualization import dataframe_visualizer
+
+
+_ALL_VISUALIZERS = [
+    dataframe_visualizer.DataframeVisualizer(),
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/as_dataframe.py:32]: In this case, as we only expect a single visualizer, it seems overkill to add this extra class. The code could go directly in `as_dataframe` (Not use `visualizer.Visualizer` at all and remove `dataframe_visualizer.py` entirely)

File: tensorflow_datasets/core/visualization/dataframe_visualizer.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+from tensorflow_datasets.core import lazy_imports_lib
+from tensorflow_datasets.core.visualization import visualizer
+
+
+class DataframeVisualizer(visualizer.Visualizer):
+  """Visualizer for text and structured datasets."""
+
+  def match(self, ds_info: dataset_info.DatasetInfo) -> bool:
+    """See base class."""
+    ds_type = visualizer.extract_dataset_type(ds_info.full_name)
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/dataframe_visualizer.py:36]: We want to support all datasets, irrespective of their type. This also include user-defined tfds datasets. So we should not restrict this to specific categories

File: tensorflow_datasets/core/visualization/visualizer.py [hunk 4]
```diff
@@ -21,15 +21,47 @@
...
+def extract_all_keys(feature_dict):
+  """Extracts keys from features dict.
+
+  Args:
+    feature_dict: `tfds.features.FeaturesDict` from which extract keys.
+
+  Returns:
+    List of extracted keys.
+  """
+  return [k for k, f in feature_dict.items()]
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/visualizer.py:62]: Ideally, we should support nested features by flattening them, like `{'x': 123, 'y': {'a': 456, 'b': 789}}` -> `['x', 'y/a', 'y/b']`
But it's fine if it's not added inn the first version

File: tensorflow_datasets/core/visualization/as_dataframe.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+]
+
+
+def as_dataframe(
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/as_dataframe.py:36]: Could you add a `as_dataframe_test.py` too which check the output of the dataframe ? Note that you should use `tfds.testing.mock_data`, like in `visualizer_test.py`

File: tensorflow_datasets/core/visualization/dataframe_visualizer.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        should not be batched. Examples will be consumed in order until
+        rows are read or the dataset is consumed.
+      ds_info: `tfds.core.DatasetInfo` object of the dataset to visualize.
+      num_examples: `int`, number of rows in the pandas dataframe.
+
+    Returns:
+      df: The pandas dataframe.
+    """
+    # Extract the keys
+    features = visualizer.extract_all_keys(ds_info.features)
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/dataframe_visualizer.py:58]: Not all features should be extracted. For instance, we want to filter images, video, and multi-array tensors.

so `tfds.as_dataframe(*tfds.load('imagenet2012', split='test', with_info=True))` would display `'label', 'filename'` but not `'image'`

File: tensorflow_datasets/core/visualization/as_dataframe.py (hunk 1 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/as_dataframe.py:68]: Removed.

File: tensorflow_datasets/core/visualization/as_dataframe.py (hunk 2 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/as_dataframe.py:32]: Fixed.

File: tensorflow_datasets/core/visualization/dataframe_visualizer.py (hunk 3 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/dataframe_visualizer.py:36]: Fixed.

File: tensorflow_datasets/core/visualization/visualizer.py (hunk 4 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/visualizer.py:62]: Sounds good.

File: tensorflow_datasets/core/visualization/as_dataframe.py (hunk 5 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/as_dataframe.py:36]: Done.

File: tensorflow_datasets/core/visualization/dataframe_visualizer.py (hunk 6 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/dataframe_visualizer.py:84]: Fixed.

File: tensorflow_datasets/core/visualization/as_dataframe_test.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+from tensorflow_datasets.core import registered
+from tensorflow_datasets.core import visualization
+
+# Import for registration
+from tensorflow_datasets.text import anli  # pylint: disable=unused-import,g-bad-import-order
+
+
+class AsDataframeTest(testing.TestCase):
+
+  @mock.patch('pandas.DataFrame')
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/as_dataframe_test.py:35]: Why mocking panda ? We could check that the values are output of `visualization.as_dataframe` is correct if you set predefined values through `.mock_data(as_dataset=...)`.

File: tensorflow_datasets/core/visualization/as_dataframe_test.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  def test_as_dataframe(self, mock_df):
+    with testing.mock_data(num_examples=20):
+      ds, ds_info = registered.load(
+          'anli', split='train', with_info=True)
+    visualization.show_examples(ds, ds_info)
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/as_dataframe_test.py:40]: You're not testing the right funtion ?

File: tensorflow_datasets/core/visualization/as_dataframe.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+      info. Available either through `tfds.load('mnist', with_info=True)` or
+      `tfds.builder('mnist').info`
+    **options_kwargs: Additional display options, specific to the dataset type
+      to visualize. Are forwarded to `tfds.visualization.Visualizer.show`.
+      See the `tfds.visualization` for a list of available visualizers.
+
+  Returns:
+    dataframe: The `pandas.DataFrame` object
+  """
+  visualizer = dataframe_visualizer.DataframeVisualizer()
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/as_dataframe.py:61]: Let's remove `DataframeVisualizer`. This add an extra indirection which feel unnecessary. Let's remove `dataframe_visualizer.py` too annd move the code here.

File: tensorflow_datasets/core/visualization/dataframe_visualizer.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    if isinstance(feature_type, features_lib.Tensor):
+      for tensor_key in keys:
+        if len(ds_info.features[tensor_key].shape) == 0:
+          keys.remove(tensor_key)
+    for key in keys:
+      all_features.remove(key)
+  return all_features
+
+
+class DataframeVisualizer(visualizer.Visualizer):
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/dataframe_visualizer.py:57]: Let's not inherit from `visualizer.Visualizer`. Let's remove the class entirely.

File: tensorflow_datasets/core/visualization/visualizer.py [hunk 11]
```diff
@@ -21,15 +21,47 @@
...
 from typing import Any
 
 import six
 import tensorflow.compat.v2 as tf
 
+import tensorflow_datasets as tfds
 from tensorflow_datasets.core import dataset_info
 
 
+def extract_dataset_type(full_name):
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/visualizer.py:37]: Let's remove this as this isn't used anywhere

File: tensorflow_datasets/core/visualization/dataframe_visualizer.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    Returns:
+      The pandas dataframe.
+    """
+    # Extract the features
+    features = visualizer.extract_all_keys(ds_info.features)
+
+    # Filter feature types that are not supported
+    unsupported_features = [
+        features_lib.Audio, features_lib.Image, features_lib.Sequence,
+        features_lib.Tensor, features_lib.Video
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/dataframe_visualizer.py:89]: As explain previously, we should support tensor if they are scalar.

File: tensorflow_datasets/core/visualization/as_dataframe_test.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/visualization/as_dataframe_test.py:41]: Let's add a test with an image dataset (e.g. imagenet) to make sure that the return column contain the label and filename, but not the image.

**Comment by Conchylicultor** [comment]
Could you just fix the tests (adding more tests of the panda output, test with image dataset to make sure the filtering works properly,...) ? And I'll fix the rest internally when merging.

**Comment by Cut-Guyo9b** [comment]
I think i the ProductNow Crazy i read that mad at my BANK

File: tensorflow_datasets/core/visualization/as_dataframe.py (hunk 9 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/as_dataframe.py:61]: Done.

File: tensorflow_datasets/core/visualization/dataframe_visualizer.py (hunk 10 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/dataframe_visualizer.py:57]: Done.

File: tensorflow_datasets/core/visualization/visualizer.py (hunk 11 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/visualizer.py:37]: Done.

File: tensorflow_datasets/core/visualization/as_dataframe_test.py (hunk 8 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/as_dataframe_test.py:40]: Fixed.

File: tensorflow_datasets/core/visualization/as_dataframe_test.py (hunk 7 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/as_dataframe_test.py:35]: Removed.

File: tensorflow_datasets/core/visualization/as_dataframe_test.py (hunk 13 above)
Comment (sharanramjee) [reply tensorflow_datasets/core/visualization/as_dataframe_test.py:41]: Added.

**Comment by vijayphoenix** [comment]
Merged by https://github.com/tensorflow/datasets/commit/6e2540f85bbfd0312d5611c2612cc38de98084f3

**Pull request closed without merging**